lib/auth/password.go

```diff
@@ -453,55 +453,28 @@ func (s *Server) changeUserSecondFactor(req *proto.ChangeUserAuthenticationReque
 	if secondFactor == constants.SecondFactorOff {
 		return nil
 	}
-	switch req.NewMFARegisterResponse.GetResponse().(type) {
-	case *proto.MFARegisterResponse_TOTP:
-		if secondFactor == constants.SecondFactorU2F {
-			return trace.BadParameter("user %q sent an OTP token during password reset but cluster only allows U2F for second factor", username)
-		}
-		secrets, err := s.Identity.GetUserTokenSecrets(ctx, req.TokenID)
-		if err != nil {
-			return trace.Wrap(err)
-		}
 
-		dev, err := services.NewTOTPDevice("otp", secrets.GetOTPKey(), s.clock.Now())
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if err := s.checkTOTP(ctx, username, req.GetNewMFARegisterResponse().GetTOTP().GetCode(), dev); err != nil {
-			return trace.Wrap(err)
-		}
-		if err := s.UpsertMFADevice(ctx, username, dev); err != nil {
-			return trace.Wrap(err)
+	if req.GetNewMFARegisterResponse() == nil {
+		switch secondFactor {
+		case constants.SecondFactorOptional:
+			return nil
```

Comment:
I think what usually happens for optional is that, if they have a device registered, then it should be used (but the user isn't forced to register a device).

Please leave a TODO to check for a device, or a comment stating why it's not necessary to check here.